File: app/src/main/java/indi/yume/app/passwordbox/util/GeneratePwd.java
package indi.yume.app.passwordbox.util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import rx.functions.Func0;

/**
 * Created by bush2 on 2016/4/4.
 */
public class GeneratePwd {
    private static final Random random = new Random(0);

    public static String getPwd(int length,
                                boolean hasBigger,
                                boolean hasSmall,
                                boolean hasNumber,
                                boolean hasSpecial) {
        int size = 0;
        if(hasBigger) size++;
        if(hasSmall) size++;
        if(hasNumber) size++;
        if(hasSpecial) size++;
        if(size == 0) {
            hasSmall = true;
            size = 1;
        }

        List<Integer> list = getRandomList(length - size, size);
        List<Integer> randomNumList = new ArrayList<>();
        int index = 0;
        if(hasBigger) {
            randomNumList.add(list.get(index) + 1);
            index++;
        } else {
            randomNumList.add(0);
        }
        if(hasSmall) {
            randomNumList.add(list.get(index) + 1);
            index++;
        } else {
            randomNumList.add(0);
        }
        if(hasNumber) {
            randomNumList.add(list.get(index) + 1);
            index++;
        } else {
            randomNumList.add(0);
        }
        if(hasSpecial) {
            randomNumList.add(list.get(index) + 1);
            index++;
        } else {
            randomNumList.add(0);
        }

        return getPwd(randomNumList.get(0),
                randomNumList.get(1),
                randomNumList.get(2),
                randomNumList.get(3));
    }

    public static List<Integer> getRandomList(int sum, int size) {
        if(size <= 0)
            return new ArrayList<>();
        if(size == 1) {
            List<Integer> list = new ArrayList<>();
            list.add(sum);
            return list;
        }

        List<Integer> randomList = new ArrayList<>();
        sum = Math.max(sum, size);

        for(int i = 0; i < size - 1; i++)
            randomList.add(sum <= 0 ? 0 : random.nextInt(sum));

        Collections.sort(randomList);

        List<Integer> list = new ArrayList<>();
        int lastNum = 0;
        int nextNum = randomList.get(0);
        for(int i = 1; i < size; i++) {
            list.add(nextNum - lastNum);
            if(i >= randomList.size())
                break;
            lastNum = nextNum;
            nextNum = randomList.get(i);
        }
        list.add(sum - nextNum);

        return list;
    }

    public static String getPwd(int biggerAtoZNum,
                                int smallAtoZNum,
                                int numberNum,
                                int specialNum) {
        StringBuilder pwd = new StringBuilder(biggerAtoZNum + smallAtoZNum + numberNum + specialNum);
        pwd.append(getString(biggerAtoZNum, GeneratePwd::getBiggerAtoZ))
                .append(getString(smallAtoZNum, GeneratePwd::getSmallAtoZ))
                .append(getString(numberNum, GeneratePwd::getNumber))
                .append(getString(specialNum, GeneratePwd::getSpecial));

        List<Character> pwdList = new ArrayList<>();
        for(char c : pwd.toString().toCharArray())
            pwdList.add(c);

        Collections.shuffle(pwdList, random);
        StringBuilder pwdBuilder = new StringBuilder();
        for(Character c : pwdList)
            pwdBuilder.append(c);
        return pwdBuilder.toString();
    }

    private static String getString(int length, Func0<Character> getChar) {
        StringBuilder randomString = new StringBuilder(length);
        for(int i = 0; i < length; i++)
            randomString.append(getChar.call());

        return randomString.toString();
    }

    public static char getSmallAtoZ() {
        return (char) ('a' + nextAbsInt() % ('z' - 'a'));
    }

    public static char getBiggerAtoZ() {
        return (char) ('A' + nextAbsInt() % ('Z' - 'A'));
    }

    public static char getNumber() {
        return (char) (random.nextInt(10) + '0');
    }

    public static char getSpecial() {
        return Constants.SPECIAL_CHARS[random.nextInt(Constants.SPECIAL_CHARS.length)];
    }

    private static int nextAbsInt() {
        return Math.abs(random.nextInt());
    }
}

File: appkotlin/src/main/kotlin/indi/yume/view/appkotlin/util/GeneratePwd.kt
package indi.yume.view.appkotlin.util

import java.util.ArrayList
import java.util.Collections
import java.util.Random

import rx.functions.Func0

/**
 * Created by bush2 on 2016/4/4.
 */
object GeneratePwd {
    private val random = Random(0)

    fun getPwd(length: Int,
               hasBigger: Boolean,
               hasSmall: Boolean,
               hasNumber: Boolean,
               hasSpecial: Boolean): String {
        var hasSmall = hasSmall
        var size = 0
        if (hasBigger) size++
        if (hasSmall) size++
        if (hasNumber) size++
        if (hasSpecial) size++
        if (size == 0) {
            hasSmall = true
            size = 1
        }

        val list = getRandomList(length - size, size)
        val randomNumList = ArrayList<Int>()
        var index = 0
        if (hasBigger) {
            randomNumList.add(list[index] + 1)
            index++
        } else {
            randomNumList.add(0)
        }
        if (hasSmall) {
            randomNumList.add(list[index] + 1)
            index++
        } else {
            randomNumList.add(0)
        }
        if (hasNumber) {
            randomNumList.add(list[index] + 1)
            index++
        } else {
            randomNumList.add(0)
        }
        if (hasSpecial) {
            randomNumList.add(list[index] + 1)
            index++
        } else {
            randomNumList.add(0)
        }

        return getPwd(randomNumList[0],
                randomNumList[1],
                randomNumList[2],
                randomNumList[3])
    }

    fun getRandomList(sum: Int, size: Int): List<Int> {
        if (size <= 0)
            return ArrayList()
        if (size == 1)
            return listOf(sum)

        val randomList = ArrayList<Int>()
        val sumItemNum = Math.max(sum, size)

        for (i in 0..size - 2)
            randomList.add(if (sumItemNum <= 0) 0 else random.nextInt(sumItemNum))

        Collections.sort(randomList)

        val list = ArrayList<Int>()
        var lastNum = 0
        var nextNum = randomList[0]
        for (i in 1..size - 1) {
            list += nextNum - lastNum
            if (i >= randomList.size)
                break
            lastNum = nextNum
            nextNum = randomList[i]
        }
        list += sumItemNum - nextNum

        return list
    }

    fun getPwd(biggerAtoZNum: Int,
               smallAtoZNum: Int,
               numberNum: Int,
               specialNum: Int): String {
        val pwd = StringBuilder(biggerAtoZNum + smallAtoZNum + numberNum + specialNum)
        pwd.apply {
            putChar(biggerAtoZNum) { biggerAtoZ }
            putChar(smallAtoZNum)  { smallAtoZ }
            putChar(numberNum)     { number }
            putChar(specialNum)    { special }
        }

        val pwdList = pwd.toList()
        Collections.shuffle(pwdList, random)
        return pwdList.joinToString(separator = "")
    }

    private fun StringBuilder.putChar(length: Int, getChar: () -> Char) =
            (0..length - 1).forEach { append(getChar()) }

    val smallAtoZ: Char
        get() = ('a' + nextAbsInt() % ('z' - 'a')).toChar()

    val biggerAtoZ: Char
        get() = ('A' + nextAbsInt() % ('Z' - 'A')).toChar()

    val number: Char
        get() = (random.nextInt(10) + '0'.toInt()).toChar()

    val special: Char
        get() = Constants.SPECIAL_CHARS[random.nextInt(Constants.SPECIAL_CHARS.size)]

    private fun nextAbsInt() = Math.abs(random.nextInt())
}
